File: src/services/supabase/promotionSyncService.js
import { appService } from '../appService.js';

const TABLE = 'promotions';

export class PromotionSyncService {
  constructor(db) {
    this._db = db;
    this._channel = null;
  }

  get _supabase() { return appService.supabaseClient; }

  async syncToSupabase() {
    let uploaded = 0, failed = 0;
    try {
      const unsynced = await this._db.promotions.getUnsynced();
      if (!unsynced.length) return { uploaded: 0, downloaded: 0, failed: 0 };

      for (const promo of unsynced) {
        try {
          const { error } = await this._supabase.from(TABLE).upsert(this._toJson(promo));
          if (error) throw error;
          uploaded++;
        } catch (e) {
          console.error('Failed to sync promotion ' + promo.id + ':', e);
          failed++;
        }
      }

      if (uploaded > 0) {
        const syncedIds = unsynced.slice(0, uploaded).map(p => p.id);
        await this._db.promotions.markSynced(syncedIds);
      }
    } catch (e) {
      console.error('Sync to Supabase error:', e);
    }
    return { uploaded, downloaded: 0, failed };
  }

  async syncFromSupabase({ since } = {}) {
    let downloaded = 0;
    try {
      let query = this._supabase.from(TABLE).select();
      if (since) query = query.gte('updated_at', since instanceof Date ? since.toISOString() : since);

      const { data, error } = await query;
      if (error) throw error;

      await this._db.promotions.bulkInsert(data ?? []);
      downloaded = (data ?? []).length;
    } catch (e) {
      console.error('Sync from Supabase error:', e);
    }
    return { uploaded: 0, downloaded, failed: 0 };
  }

  async fullSync() {
    const uploadResult = await this.syncToSupabase();
    const downloadResult = await this.syncFromSupabase();
    return {
      uploaded: uploadResult.uploaded,
      downloaded: downloadResult.downloaded,
      failed: uploadResult.failed,
    };
  }

  async deleteFromSupabase(id) {
    try {
      const { error } = await this._supabase.from(TABLE).delete().eq('id', id);
      if (error) throw error;
      return true;
    } catch (e) {
      console.error('Delete from Supabase error:', e);
      return false;
    }
  }

  setupRealtimeSync({ onInsert, onUpdate, onDelete }) {
    this._channel?.unsubscribe?.();
    this._channel = this._supabase
      .channel('promotions_changes')
      .on('postgres_changes', { event: '*', schema: 'public', table: TABLE }, async (payload) => {
        try {
          if (payload.eventType === 'INSERT') {
            await this._db.promotions.insert(payload.new);
            onInsert?.(payload.new);
          } else if (payload.eventType === 'UPDATE') {
            await this._db.promotions.updatePromotion(payload.new);
            onUpdate?.(payload.new);
          } else if (payload.eventType === 'DELETE') {
            const id = payload.old?.id;
            if (id != null) {
              await this._db.promotions.deletePromotion(id);
              onDelete?.(id);
            }
          }
        } catch (e) {
          console.error('Realtime promotion error:', e);
        }
      })
      .subscribe();
  }

  disposeRealtimeSync() {
    this._channel?.unsubscribe?.();
    this._channel = null;
  }

  _toJson(promo) {
    return {
      id: promo.id,
      name: promo.name,
      description: promo.description,
      type: promo.type,
      applicable_product_ids: Array.isArray(promo.applicable_product_ids)
        ? promo.applicable_product_ids.join(',')
        : (promo.applicable_product_ids ?? ''),
      min_quantity: promo.min_quantity,
      discount_percent: promo.discount_percent,
      discount_amount: promo.discount_amount,
      free_product_id: promo.free_product_id,
      free_product_qty: promo.free_product_qty,
      discount_on_items: promo.discount_on_items,
      is_active: promo.is_active,
      start_date: promo.start_date,
      end_date: promo.end_date,
      priority: promo.priority,
      created_at: promo.created_at,
      updated_at: promo.updated_at,
    };
  }
}

File: services/promotions/promotion_sync_service.dart
import 'package:hse_pos/app/core/locators/service_locator.dart';
import 'package:hse_pos/app/databases/hsepos_database.dart';
import 'package:hse_pos/app/models/promotion_model/promotion_model.dart';
import 'package:hse_pos/app/repos/promotion_repository/promotion_repository.dart';
import 'package:hse_pos/app/services/promotions/promotion_service.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

class PromotionSyncService {
  static const String _tableName = 'promotions';
  static SupabaseClient get _client => Supabase.instance.client;

  final PromotionRepository promotionDatabase = PromotionRepository(
    locator<TestDatabase>(),
  );

  /// Sync local to Supabase
  Future<SyncResult> syncToSupabase() async {
    int uploaded = 0;
    int failed = 0;

    try {
      final unsynced = await promotionDatabase.getUnsynced();
      if (unsynced.isEmpty) {
        return SyncResult(uploaded: 0, downloaded: 0, failed: 0);
      }

      for (final promo in unsynced) {
        try {
          await _client.from(_tableName).upsert(promo.toJson());
          uploaded++;
        } catch (e) {
          print('Failed to sync ${promo.id}: $e');
          failed++;
        }
      }

      // Mark synced
      if (uploaded > 0) {
        final syncedIds = unsynced.take(uploaded).map((p) => p.id).toList();
        await promotionDatabase.markSynced(syncedIds);
      }
    } catch (e) {
      print('Sync to Supabase error: $e');
    }

    return SyncResult(uploaded: uploaded, downloaded: 0, failed: failed);
  }

  /// Sync from Supabase to local
  Future<SyncResult> syncFromSupabase({DateTime? since}) async {
    int downloaded = 0;

    try {
      var query = _client.from(_tableName).select();

      if (since != null) {
        query = query.gte('updated_at', since.toIso8601String());
      }

      final response = await query;

      final promotions = response
          .map((json) => PromotionModel.fromJson(json))
          .toList();

      await promotionDatabase.bulkInsert(promotions);
      downloaded = promotions.length;

      // Clear cache after sync
      PromotionService.clearCache();
    } catch (e) {
      print('Sync from Supabase error: $e');
    }

    return SyncResult(uploaded: 0, downloaded: downloaded, failed: 0);
  }

  /// Full two-way sync
  Future<SyncResult> fullSync() async {
    // First push local changes
    final uploadResult = await syncToSupabase();

    // Then pull from Supabase
    final downloadResult = await syncFromSupabase();

    return SyncResult(
      uploaded: uploadResult.uploaded,
      downloaded: downloadResult.downloaded,
      failed: uploadResult.failed,
    );
  }

  /// Delete from Supabase
  static Future<bool> deleteFromSupabase(String id) async {
    try {
      await _client.from(_tableName).delete().eq('id', id);
      return true;
    } catch (e) {
      print('Delete from Supabase error: $e');
      return false;
    }
  }

  /// Setup realtime listener
  static RealtimeChannel? _channel;

  void setupRealtimeSync({
    required Function(PromotionModel) onInsert,
    required Function(PromotionModel) onUpdate,
    required Function(String) onDelete,
  }) {
    _channel?.unsubscribe();

    _channel = _client
        .channel('promotions_changes')
        .onPostgresChanges(
          event: PostgresChangeEvent.all,
          schema: 'public',
          table: _tableName,
          callback: (payload) async {
            switch (payload.eventType) {
              case PostgresChangeEvent.insert:
                final promo = PromotionModel.fromJson(payload.newRecord);
                await promotionDatabase.insert(promo);
                onInsert(promo);
                break;
              case PostgresChangeEvent.update:
                final promo = PromotionModel.fromJson(payload.newRecord);
                promotionDatabase.updatePromotion(promo);
                onUpdate(promo);
                break;
              case PostgresChangeEvent.delete:
                final id = payload.oldRecord['id'] as String?;
                if (id != null) {
                  promotionDatabase.deletePromotion(id);
                  onDelete(id);
                }
                break;
              default:
                break;
            }
            PromotionService.clearCache();
          },
        )
        .subscribe();
  }

  static void disposeRealtimeSync() {
    _channel?.unsubscribe();
    _channel = null;
  }
}

class SyncResult {
  final int uploaded;
  final int downloaded;
  final int failed;

  SyncResult({
    required this.uploaded,
    required this.downloaded,
    required this.failed,
  });

  @override
  String toString() =>
      'Synced: ↑$uploaded ↓$downloaded${failed > 0 ? ' ⚠️$failed failed' : ''}';
}
